spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -17,37 +17,144 @@
+2. The microprofile.io implementation must provide a mechanism to automatically extract Span information from any incoming JAX-RS request.
+3. The microprofile.io implementation must provide a mechanism to automatically start a Span for any incoming JAX-RS request.
+5. The microprofile.io implementation must provide a mechanism to automatically inject Span information into any outgoing JAX-RS request.
+4. The microprofile.io implementation must provide a mechanism to automatically start a Span for any outgoing JAX-RS request.
 
-This proposal also specifies that the underlying opentracing.io Tracer object configured instance is available for developer use. The microprofile.io implementation will make the configured Tracer available with CDI injection.
+All automatically created Spans must have the correct parent/child relationships when creating either synchronous or asynchronous JAX-RS incoming and outgoing requests.
 
-The proposal also specifies two behaviors associated with JAX-RS applications, that will make it easier for applications to propagate Spans.
+==== Tracer configuration
+An implementation of an io.opentracing.Tracer must be made available to each application. Each application will have its own Tracer instance.
+The Tracer must be configurable outside of the application to match the distributed tracing environment where the application is deployed. For example, it should be possible to take the exact same application and deploy it to an environment where Zipkin is in use, and to deploy the application to a different environment where Jaeger is in use, and the application should report Spans correctly in either environment.
 
-* Span extraction
+==== Span extraction
+When a request arrives at a JAX-RS endpoint, an attempt is made to use the configured Tracer to extract a Span from the arriving request. If a Span is extracted, it is used as the parent Span for the new Span that is created for the endpoint.
 
-* Span insertion
+==== Span creation for inbound requests
+When a request arrives at a JAX-RS endpoint, a new Span is created. The new Span will be a child of the Span extracted from the incoming request, if the extracted Span exists.
 
+===== Span name
+The default operation name of the new Span for the incoming request is
+```
+<HTTP method>:<package name>.<Class name>.<method name>
+```
 
+===== Span tags
+Spans created for incoming requests will have the following tags added by default:
+
+* Tags.SPAN_KIND = Server
+* Tags.HTTP_METHOD
+* Tags.HTTP_URL
+* Tags.HTTP_STATUS
+* Tags.ERROR (if true)
+
+==== Span creation and injection for outbound requests
+When a request is sent from a JAX-RS  javax.ws.rs.client.Client, a new Span will be created that is injected in the outbound request for propagation downstream. The new Span will be a child of the current Span if a current Span exists. The new Span will be finished when the outbound request is completed.
```

Comment:
Some http client instrumentations go a little bit beyond this:

One span which wraps the whole invocation and N spans which are descendants of the first span to model each network attempt (e.g. redirects..). It's a little bit more verbose, but provides a better understanding.

Reply:
Can you point me to a particular client implementation to look at?